### PR TITLE
DISPATCH-1372 alloc_pool intrusive linked list can be replaced by a linked stack

### PR DESCRIPTION
**Pull request by franz1981**

### CODE REVIEW

**Comment by franz1981** [comment]
Please do not merge it yet: it has been sent for review :+1:

File: src/alloc_pool.c [hunk 1]
```diff
@@ -44,20 +46,97 @@ DEQ_DECLARE(qd_alloc_type_t, qd_alloc_type_list_t);
...
+struct qd_alloc_chunk_t {
+    qd_alloc_chunk_t     *prev;
+    qd_alloc_item_t      *items[CHUNK_SIZE];
+    qd_alloc_chunk_t     *next;
+};
+
+struct qd_alloc_linked_stack_t {
+    //the base
+    qd_alloc_chunk_t     *top_chunk;
+    uint32_t              top;          //qd_alloc_item* top_item = top_chunk->items[top+1] <-> top > 0
```
Comment (astitcher) [inline src/alloc_pool.c]: Probably want to reorder `top` and `size`. On 64 bit machine there would be 32 bits of padding in the middle of the struct with this ordering.

Comment (astitcher) [reply src/alloc_pool.c]: Actually as `qp_alloc_chunk_t` contains only pointers you probably want to put `size` as the very last member in this structure (but you'll still get some padding)

Comment (franz1981) [reply src/alloc_pool.c]: We can try to reorder fields and see if it will get better cache localty: I see that `qd_alloc_linked_stack_t` is suboptimal, but I've just tried to put `base_chunk` near to the most accessed fields to make used of the adiacent line prefetching of many Intel CPUs; so I've focused just on that...
I can see too that `CAPACITY` could be changed to improve things too...

File: src/alloc_pool.c [hunk 2]
```diff
@@ -44,20 +46,97 @@ DEQ_DECLARE(qd_alloc_type_t, qd_alloc_type_list_t);
...
-    DEQ_LINKS(qd_alloc_item_t);
     uint32_t              sequence;
 #ifdef QD_MEMORY_DEBUG
     qd_alloc_type_desc_t *desc;
     uint32_t              header;
 #endif
 };
 
-DEQ_DECLARE(qd_alloc_item_t, qd_alloc_item_list_t);
+#define CHUNK_SIZE 128/sizeof(void*)
```
[the remark this answers is not in the document]
Comment (franz1981) [reply src/alloc_pool.c]: Uh right, I haven't commented it: it is using as the base size see the Adjacent Cache-Line Prefetch size ie 128 bytes (https://software.intel.com/en-us/articles/optimizing-application-performance-on-intel-coret-microarchitecture-using-hardware-implemented-prefetchers)

File: src/alloc_pool.c [hunk 3]
```diff
@@ -44,20 +46,97 @@ DEQ_DECLARE(qd_alloc_type_t, qd_alloc_type_list_t);
...
 
+struct qd_alloc_chunk_t {
+    qd_alloc_chunk_t     *prev;
```
[the remark this answers is not in the document]
Comment (franz1981) [reply src/alloc_pool.c]: I wanted prev/next to be in that specific order: I know that is not 100% guaranteed by many compilers, but I see that clang and gcc tends to stick to what is written in the code

File: src/alloc_pool.c [hunk 4]
```diff
@@ -44,20 +46,97 @@ DEQ_DECLARE(qd_alloc_type_t, qd_alloc_type_list_t);
...
+    qd_alloc_item_t      *items[CHUNK_SIZE];
+    qd_alloc_chunk_t     *next;
+};
+
+struct qd_alloc_linked_stack_t {
+    //the base
+    qd_alloc_chunk_t     *top_chunk;
+    uint32_t              top;          //qd_alloc_item* top_item = top_chunk->items[top+1] <-> top > 0
+    uint64_t              size;
+    qd_alloc_chunk_t      base_chunk;
```
[the remark this answers is not in the document]
Comment (franz1981) [reply src/alloc_pool.c]: I wanted its allocation to be part of `qd_alloc_linked_stack_t` to increase localty or am I misunderstood the question?

File: src/alloc_pool.c (hunk 1 above)
Comment (astitcher) [reply src/alloc_pool.c]: In this case I was just thinking about eliminating useless struct padding - which might allow you to fit more into a cache line, but might not help locality or prefetching.

File: src/alloc_pool.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (astitcher) [reply src/alloc_pool.c]: The C standard guarantees that members in a struct will appear in memory in the order you specify - The only change the compiler can make is to pad struct members to maintain machine dependent alignment constraints.

[the remark this answers is not in the document]
Comment (franz1981) [reply src/alloc_pool.c]: I would love to have a similar behaviour on Java :D Nice to hear it, thnks!

File: src/alloc_pool.c (hunk 1 above)
Comment (franz1981) [reply src/alloc_pool.c]: I'm not aware of any struct padding here...but maybe there are some introduced automagically somehwere?

Comment (astitcher) [reply src/alloc_pool.c]: On a 64 bit machine you'll get padding after the uint32 (`top`) to pad it out to 64 bits. Everything else in the struct is 64 bits (pointers) so to avoid the padding in the middle of the struct you want to put `top` at the end of the struct.
[On a 32 bit machine this is not a concern, but don't care as much about them anyway].

Comment (franz1981) [reply src/alloc_pool.c]: > to avoid the padding in the middle of the struct you want to put top at the end of the struct.

I haven't considered field padding, you're right!
In this case, considering that `base_chunk` is > 128 bytes I won't risk to get a cache miss on each access?

File: src/alloc_pool.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (kgiusti) [reply src/alloc_pool.c]: That answers my question - thanks.

File: src/alloc_pool.c [hunk 5]
```diff
@@ -164,14 +242,17 @@ void *qd_alloc(qd_alloc_type_desc_t *desc, qd_alloc_pool_t **tpool)
...
+            push_stack(&pool->free_list, item);
         }
     } else {
         //
         // Allocate a full batch from the heap and put it on the thread list.
         //
+        //TODO(franz):
+        //  -   would be better to allocate in batches == transfer_batch_size
+        //      and put a small (== sizeof(transfer_batch_size)) ref_count to help the final free
+        //  -   could be beneficial directly to delink a chunk?
```
[the remark this answers is not in the document]
Comment (franz1981) [reply src/alloc_pool.c]: I need to explain it better IMHO. Let me rephrase something that I haven't clarified yet on the JIRA.
The behaviour of this chunked stack is not actually the same of the original one, in particular while "rebalancing" resources into/from the global free list.
I have chosen to just move `qd_alloc_item` between free lists, but not `qd_alloc_chunk_t` on purpose, because:

- the purpose of `qd_alloc_chunk_t` is to cope with the allocation rate of `qd_alloc_item` on a same thread
- to reduce the need to move "big" chunks  between threads (causing additional cache misses on the first attempt to access them)
- thread local lists are nearly bounded in size `localFreeListMax` ie can't grow unbounded (maybe)

I'm not sure I've chosen right, that's the reason of the second comment ie 
> could be beneficial directly to delink a chunk?

Maybe! I need a paper and a pencil to think about the implications of it and which benefits will bring!

File: src/alloc_pool.c (hunk 1 above)
Comment (astitcher) [reply src/alloc_pool.c]: Hmm, difficult to say - in any event it's going to be close to some entries in the list and far from others. Perhaps eliminating it completely might be possible.

Another thing to think about is not to use pointers at all, but rather use a base (64 bit address) and smaller offset (32 bit entry count) - that should allow more entries into the same number of cache lines - you would have to allocate larger and contiguous blocks though.

Comment (franz1981) [reply src/alloc_pool.c]: I love the "compressed pointers" idea and given that we are mostly memory limited here it would allow to pack more pointers...something that using an intrusive pointer solution was really tricky (if not impossible) to implemente...

File: src/alloc_pool.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (kgiusti) [reply src/alloc_pool.c]: Franz - can you add a comment explaining that?  Otherwise someone may "clean up" that using DEQ (like myself) and not realize the effect.

File: src/alloc_pool.c [hunk 6]
```diff
@@ -164,14 +242,17 @@ void *qd_alloc(qd_alloc_type_desc_t *desc, qd_alloc_pool_t **tpool)
         desc->stats->held_by_threads += desc->config->transfer_batch_size;
 #endif
         for (idx = 0; idx < desc->config->transfer_batch_size; idx++) {
-            item = DEQ_HEAD(desc->global_pool->free_list);
-            DEQ_REMOVE_HEAD(desc->global_pool->free_list);
-            DEQ_INSERT_TAIL(pool->free_list, item);
+            item = pop_stack(&desc->global_pool->free_list);
```
[the remark this answers is not in the document]
Comment (kgiusti) [reply src/alloc_pool.c]: I'm OK with doing the chunk_size changes is a follow on patch if you can't spend time on it right now.

File: src/alloc_pool.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (franz1981) [reply src/alloc_pool.c]: Sure, and I can add a configuration option on the chunk size as well. I think (but will need to test it with Mick) that using chunks as big as the transfer size (or OS page size) will improve things a lil more too

File: src/alloc_pool.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (franz1981) [reply src/alloc_pool.c]: The tranfer in batch will require me some more time, but the change on chunk size would be pretty easy to be achieved, so I can do it

[the remark this answers is not in the document]
Comment (franz1981) [reply src/alloc_pool.c]: I've added another commit (that I would squash/drop if necessary) to test it: will work on the documentation part too. I've set the CHUNK_SIZE = 128 that seems the default `local_free_list_max` just to see if it makes any difference

[the remark this answers is not in the document]
Comment (franz1981) [reply src/alloc_pool.c]: OK it seems that the original chunk size was good enough, no significant improvements

[the remark this answers is not in the document]
Comment (kgiusti) [reply src/alloc_pool.c]: Thanks for trying that out - let's keep with the original CHUNK_SIZE then.

**Comment by kgiusti** [comment]
Just for completeness - mgoulish did some benchmarking with this patch.  Here is a summary of his results:

RESULTS  (mean latency over 1e6 messages ) (in msec)

pre-Franz:
  64.27
  64.22
  62.62
average of 3 tests : 63.70

post-Franz:
  48.37
  48.89
  50.83
average of 3 tests : 49.36


Post-Franz messages fly 1.29 times faster.
( Get from sender to receiver in 0.77 as much time. )

**Comment by franz1981** [comment]
I've added a new commit to attempt batch unordered copy between stacks: I will wait the @mgoulish response on this and If it delivers some improvement will squash and ask 4 reviews :+1:

File: src/alloc_pool.c [hunk 7]
```diff
@@ -44,20 +46,142 @@ DEQ_DECLARE(qd_alloc_type_t, qd_alloc_type_list_t);
...
 #endif
 };
 
-DEQ_DECLARE(qd_alloc_item_t, qd_alloc_item_list_t);
+//128 has been chosen because many CPUs arch use an
+//adiacent line prefetching optimization that load
+//2*cache line bytes in batch
+#define CHUNK_SIZE 128/sizeof(void*)
 
+struct qd_alloc_chunk_t {
```
Comment (ganeshmurthy) [inline src/alloc_pool.c]: So assuming sizeof(void*) in a 64 bit machine is 8, 128/8 = 16
qd_alloc_chunk_t will have 
1. 8 bytes for prev
2. 16 items (qd_alloc_item_ts)
3. 8 bytes for next
======================
Total = 128 + 16 = 144 bytes. 
=======================
But Adjacent Cache-Line Prefetch size ie 128 bytes. Looks like 144 bytes will not fit into one Adjacent Cache-Line Prefetch. So, does it make sense to move "items" as the first member of qd_alloc_chunk_t in which case we could use the DEQ for prev and next

Comment (franz1981) [reply src/alloc_pool.c]: That's correct,  the thing is that ideally I should have used next/prev to reach the 128 bytes (ie 128 bytes should be the size of the chunk struct), but I have chosen to order the fields according to the LIFO access pattern (prev is accessed only If the first item was the last accessed) and left the adjacent line prefetcher do the dirty work for the other items not accessed yet.

Comment (ganeshmurthy) [reply src/alloc_pool.c]: define CHUNK_SIZE 128/sizeof(void*) - 2
That way, like you said, one qd_alloc_chunk_t will fit in one  Adjacent Cache-Line Prefetch

Comment (franz1981) [reply src/alloc_pool.c]: I'm gonna take a look to https://software.intel.com/sites/default/files/managed/9e/bc/64-ia-32-architectures-optimization-manual.pdf again, to be sure...let me know wdyt

**Comment by franz1981** [comment]
@kgiusti Do you know if it was the version with the first or the last commit? The second commit has introduced a memcpy

**Comment by franz1981** [comment]
@kgiusti 
I've tested with a long running multiple pairs run with 25m messages and 2 routers without getting any error on shutdown: I need some help to reproduce the failure...

**Comment by franz1981** [comment]
@kgiusti I've just squashed the commit with the batch move between stacks ie `unordered_move_stack`. 
The benchmarks shows that it delivers a small (but measurable) speedup over the original version, so IMO it worths to be used.

File: src/alloc_pool.c [hunk 8]
```diff
@@ -44,20 +46,142 @@ DEQ_DECLARE(qd_alloc_type_t, qd_alloc_type_list_t);
...
+        chunk = next;
+    }
+}
+
+static inline void next_chunk_stack(qd_alloc_linked_stack_t *const stack)
+{
+    assert(stack->top == CHUNK_SIZE);
+    qd_alloc_chunk_t *top = stack->top_chunk->next;
+    if (top == NULL) {
+        top = NEW(qd_alloc_chunk_t);
```
Comment (franz1981) [inline src/alloc_pool.c]: @astitcher @kgiusti I was thiking to allocate new chunkcs aligned to cache line size, to help `unordered_move_stack` `memcpy`s, wdyt?

File: src/alloc_pool.c [hunk 9]
```diff
@@ -44,20 +46,145 @@ DEQ_DECLARE(qd_alloc_type_t, qd_alloc_type_list_t);
...
+        next_chunk_stack(stack);
+    }
+    stack->size++;
+    stack->top_chunk->items[stack->top] = item;
+    stack->top++;
+}
+
+static inline int unordered_move_stack(qd_alloc_linked_stack_t *from, qd_alloc_linked_stack_t *to, uint32_t length)
+{
+    length = from->size < length ? from->size : length;
```
[the remark this answers is not in the document]
Comment (franz1981) [reply src/alloc_pool.c]: It never happen: i have implemented this method in a general fashion, but there ATM there is no real call site that allows it to happen. I have been lucky that the way I've implemented it has allowed too handle the case where a new chunk can't be allocated too....
